File: todo-app/src/apis/todo.ts
interface Response {
  message: string;
  data?: any;
}

const API_ENDPOINT = "https://api.aws-demo.authorizer.dev/todo";

export const getTodo = async (token: string): Promise<Response> => {
  try {
    const res = await fetch(API_ENDPOINT, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });

    const json = await res.json();

    if (res.status >= 400) {
      throw new Error(json);
    }

    return json;
  } catch (err) {
    throw err;
  }
};

export const addTodo = async (
  token: string,
  title: string
): Promise<Response> => {
  try {
    const res = await fetch(API_ENDPOINT, {
      method: "POST",
      body: JSON.stringify({
        title: title,
      }),
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });

    const json = await res.json();

    if (res.status >= 400) {
      throw new Error(json);
    }

    return json;
  } catch (err) {
    throw err;
  }
};

export const updateTodo = async (
  token: string,
  todoId: string,
  isCompleted: boolean
): Promise<Response> => {
  try {
    const res = await fetch(`${API_ENDPOINT}/${todoId}`, {
      method: "PUT",
      body: JSON.stringify({
        is_completed: isCompleted,
      }),
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });

    const json = await res.json();

    if (res.status >= 400) {
      throw new Error(json);
    }

    return json;
  } catch (err) {
    throw err;
  }
};

export const deleteTodo = async (
  token: string,
  todoId: string
): Promise<Response> => {
  try {
    const res = await fetch(`${API_ENDPOINT}/${todoId}`, {
      method: "DELETE",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });

    const json = await res.json();

    if (res.status >= 400) {
      throw new Error(json);
    }

    return json;
  } catch (err) {
    throw err;
  }
};

File: apis/controllers/todo.go
package controllers

import (
	"apis/db"
	"apis/models"
	"context"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
)

// AddTodoController, handler for todo post api
func AddTodoController(c *gin.Context) {
	context := context.Background()
	todo := &models.Todo{}

	if err := c.BindJSON(todo); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": err.Error(),
		})

		return
	}

	if strings.TrimSpace(todo.Title) == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Task title required",
		})

		return
	}

	userID, _ := c.Get("user_id")
	todo.UserID = userID.(string)

	todo, err := db.AddTodo(context, todo)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"message": err.Error(),
		})

		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "Task added successfully",
		"data":    todo,
	})
}

// UpdateTodoController, handler for todo put api
func UpdateTodoController(c *gin.Context) {
	context := c.Request.Context()
	todoID := c.Param("id")

	if strings.TrimSpace(todoID) == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "todo identifier required in url parameter for this request",
		})

		return
	}
	todo, err := db.GetTodoById(context, todoID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": err.Error(),
		})
	}

	newTodo := &models.Todo{}
	if err := c.BindJSON(newTodo); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": err.Error(),
		})

		return
	}

	if newTodo.Title != "" && newTodo.Title != todo.Title {
		todo.Title = newTodo.Title
	}

	if newTodo.IsCompleted != todo.IsCompleted {
		todo.IsCompleted = newTodo.IsCompleted
	}

	todo, err = db.UpdateTodo(context, todo)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"message": err.Error(),
		})

		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "Task updated successfully",
		"data":    todo,
	})
}

// DeleteTodoController, handler for todo delete api
func DeleteTodoController(c *gin.Context) {
	context := c.Request.Context()
	todoID := c.Param("id")

	if strings.TrimSpace(todoID) == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "todo identifier required in url parameter for this request",
		})

		return
	}
	_, err := db.GetTodoById(context, todoID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": err.Error(),
		})
	}

	err = db.DeleteTodo(context, todoID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"message": err.Error(),
		})

		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "Task deleted successfully",
	})
}

// GetTodoController, handler to get todo as per userID
func GetTodoController(c *gin.Context) {
	context := c.Request.Context()
	userID, _ := c.Get("user_id")

	todos, err := db.GetTodos(context, userID.(string))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "Tasks listed successfully",
		"data":    todos,
	})
}
